### ISSUE TITLE
`#[expect(...)]` attribute not working for `clippy::unsafe_derive_deserialize`

### ISSUE THREAD

**Issue by B14CK313** [rust-lang/rust-clippy#12802]
### Summary

The `#[expect(clippy::unsafe_derive_deserialize)]` attribute (enabled by the [`lint_reasons` feature](https://github.com/rust-lang/rust/issues/85549)) doesn't suppress the warning about deriving `serde::Deserialize` on a struct with an `unsafe fn`.
At the same time the `expect` attribute gives a warning about the lint expectation being unfulfilled.

At first glance this looked like the same bug as [rust#119358](https://github.com/rust-lang/rust/issues/119358) but in that case the warning got suppressed, which isn't the case here.

If this issue is not caused by clippy, I am happy to copy this issue over to the rust repo.

### Reproducer

I tried this code:

```rust
#![feature(lint_reasons)]
#![warn(clippy::unsafe_derive_deserialize)]

#[expect(clippy::unsafe_derive_deserialize)]
#[derive(serde::Deserialize)]
struct Foo {}

impl Foo {
    #[allow(dead_code)]
    unsafe fn unsafe_fn(&self) {
        todo!()
    }
}
```

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=093374712571828de77044e237d674e3)

I expected to see this happen:
No warning emitted.

Instead, this happened:
Two warnings get emitted:
- ``warning: you are deriving `serde::Deserialize` on a type that has methods using `unsafe` ``
- ``warning: this lint expectation is unfulfilled``

### Version

```text
rustc: 1.80.0-nightly (2024-05-14 8387315ab3c26a57a1f5)
clippy: 0.1.80 (2024-05-14 8387315)
```
